### PR TITLE
[dbnode] Refactoring dbShard

### PR DESCRIPTION
**Pull request by linasm**
**What this PR does / why we need it**:
Refactors `dbShard`, aligns with the new linter, removes some dead code.

**Special notes for your reviewer**:

**Does this PR introduce a user-facing and/or backwards incompatible change?**:
NONE

**Does this PR require updating code package or user-facing documentation?**:
NONE

### CODE REVIEW

File: src/dbnode/storage/namespace.go [hunk 1]
```diff
@@ -1778,7 +1774,7 @@ func (n *dbNamespace) aggregateTiles(
 		}
 
 		shardProcessedTileCount, err := targetShard.AggregateTiles(
-			sourceNs.ID(), sourceShard.ID(), blockReaders, writer, sourceBlockVolumes, opts, nsCtx.Schema)
+			sourceNs.ID(), n, sourceShard.ID(), blockReaders, writer, sourceBlockVolumes, opts)
```
Comment (arnikola) [inline src/dbnode/storage/namespace.go:1777]: Should this be done under RLock if we're passing in `n`?

File: src/dbnode/storage/shard.go [hunk 2]
```diff
@@ -2663,21 +2659,24 @@ func (s *dbShard) Repair(
...
+			"blockReaders and sourceBlockVolumes length mismatch (%d != %d)",
+			len(blockReaders),
+			len(sourceBlockVolumes))
 	}
 
 	openBlockReaders := make([]fs.DataFileSetReader, 0, len(blockReaders))
 	defer func() {
 		for _, reader := range openBlockReaders {
-			reader.Close()
+			_ = reader.Close()
```
Comment (arnikola) [inline src/dbnode/storage/shard.go:2679]: nit; may be worth logging these? If they'd be too noisy on errors here, add a comment to that effect

File: src/dbnode/storage/shard.go [hunk 3]
```diff
@@ -2764,54 +2732,11 @@ func (s *dbShard) AggregateTiles(
...
-		}
-
-		segment.Finalize()
-	}
+	var multiErr xerrors.MultiError
 
-	if err := readerIter.Err(); err != nil {
+	processedTileCount, err := s.tileAggregator.AggregateTiles(
+		opts, targetNs, s.ID(), openBlockReaders, writer)
+	if err != nil {
```
Comment (arnikola) [inline src/dbnode/storage/shard.go:2739]: nit: add a comment here that we still need to do cleanup etc and should log any errors from that

File: src/dbnode/storage/shard_test.go [hunk 4]
```diff
@@ -1908,21 +1877,8 @@ func TestShardAggregateTiles(t *testing.T) {
...
-		dp(secondSourceBlockStart.Add(2*time.Minute), 5+3),
-		dp(secondSourceBlockStart.Add(3*time.Minute), 0),
-		dp(secondSourceBlockStart.Add(4*time.Minute), 9),
-	}, xtime.Nanosecond, counterPayload)
-
-	write3 := newWrittenDataMatcher(t, []ts.Datapoint{
-		dp(secondSourceBlockStart.Add(time.Minute), 3),
-	}, xtime.Nanosecond, gaugePayload)
+	targetShard := testDatabaseShardWithIndexFn(t, testOpts, nil, true)
+	defer targetShard.Close() // nolint:errcheck
```
Comment (arnikola) [inline src/dbnode/storage/shard_test.go:1881]: nit; may be worth doing a 
```golang
defer func() {
 assert.NoError(t, targetShard.Close())
}()
```
here instead to ensure closing is always clean?

File: src/dbnode/storage/shard_test.go [hunk 5]
```diff
@@ -1932,16 +1888,18 @@ func TestShardAggregateTiles(t *testing.T) {
...
-		writer.EXPECT().WriteAll(id1, tags1, write1, gomock.Any()),
-		writer.EXPECT().WriteAll(id2, tags2, write2, gomock.Any()),
-		writer.EXPECT().WriteAll(id3, tags3, write3, gomock.Any()),
 		writer.EXPECT().Close(),
 	)
 
+	targetNs := NewMockNamespace(ctrl)
+	aggregator.EXPECT().
+		AggregateTiles(opts, targetNs, sourceShard.ID(), gomock.Len(2), writer).
+		Return(int64(4), nil)
```
Comment (arnikola) [inline src/dbnode/storage/shard_test.go:1899]: nit: can this 4 be a named variable to give it some context?

File: src/dbnode/storage/shard_test.go [hunk 6]
```diff
@@ -1932,16 +1888,18 @@ func TestShardAggregateTiles(t *testing.T) {
 			BlockStart:          opts.Start,
 			BlockSize:           targetBlockSize,
 			VolumeIndex:         1,
-			PlannedRecordsCount: 2,
+			PlannedRecordsCount: 3,
```
Comment (arnikola) [inline src/dbnode/storage/shard_test.go:1891]: Just wondering why this has increased?

**Review by arnikola — APPROVED** [review]
LGTM with a few nits

File: src/dbnode/storage/namespace.go (hunk 1 above)
Comment (linasm) [reply src/dbnode/storage/namespace.go:1777]: The method called accepts `n` as a `storage.Namespace` interface which only exposes public methods of `dbNamespace`. These public methods have locking when they access mutable fields of `dbNamespace` (eg. `dbNamespace.Schema()`. So I guess this should be fine?

File: src/dbnode/storage/shard.go (hunk 2 above)
Comment (linasm) [reply src/dbnode/storage/shard.go:2679]: Will add logging, should not be noisy at all.

File: src/dbnode/storage/shard.go (hunk 3 above)
Comment (linasm) [reply src/dbnode/storage/shard.go:2739]: 👍

File: src/dbnode/storage/shard_test.go (hunk 4 above)
Comment (linasm) [reply src/dbnode/storage/shard_test.go:1881]: Will do. I guess just `defer assert.NoError(t, targetShard.Close())` should work.

File: src/dbnode/storage/shard_test.go (hunk 5 above)
Comment (linasm) [reply src/dbnode/storage/shard_test.go:1899]: 👍

File: src/dbnode/storage/shard_test.go (hunk 6 above)
Comment (linasm) [reply src/dbnode/storage/shard_test.go:1891]: To cover some additional logics; will make this self-documenting by introducing named vars.

File: src/dbnode/storage/shard.go [hunk 7]
```diff
@@ -2676,7 +2676,9 @@ func (s *dbShard) AggregateTiles(
 	openBlockReaders := make([]fs.DataFileSetReader, 0, len(blockReaders))
 	defer func() {
 		for _, reader := range openBlockReaders {
-			_ = reader.Close()
+			if err := reader.Close(); err != nil {
+				s.logger.Error("error closing DataFileSetReader", zap.Error(err))
```
Comment (arnikola) [inline src/dbnode/storage/shard.go:2680]: nit: typically no `error` in go error returns, change to something like "could not close DataFileSetReader"

**Review by arnikola — APPROVED** [review]

File: src/dbnode/storage/shard.go (hunk 7 above)
Comment (linasm) [reply src/dbnode/storage/shard.go:2680]: 👍

**Pull request merged**